lib/web_ui/lib/src/engine/html/clip.dart

```diff
@@ -291,26 +289,68 @@ class PersistedPhysicalShape extends PersistedContainerSurface
           if (clipBehavior != ui.Clip.none) {
             style.overflow = 'hidden';
           }
+          applyCssShadow(rootElement, pathBounds, elevation, shadowColor);
           return;
         }
       }
     }
 
-    final String svgClipPath = _pathToSvgClipPath(path,
-        offsetX: -pathBounds.left,
-        offsetY: -pathBounds.top,
-        scaleX: 1.0 / pathBounds.width,
-        scaleY: 1.0 / pathBounds.height);
-    // If apply is called multiple times (without update) , remove prior
-    // svg clip element.
+    /// If code reaches this point, we have a path we want to clip against and
+    /// potentially have a shadow due to material surface elevation.
+    ///
+    /// When there is no shadow we can simply clip a div with a background
+    /// color using an svg clip path.
+    ///
+    /// Otherwise we need to paint a svg element for the path and clip
+    /// contents against same path for shadow to work since box-shadow doesn't
+    /// take clip-path into account.
+    ///
+    /// Webkit has a bug when applying clip-path on an element that has
+    /// position: absolute and transform. To place clipping rectangle correctly
+    /// we size the inner container to cover full pathBounds instead of sizing
+    /// to clipping rect bounds (which is the case for elevation == 0.0 where
+    /// we shift outer/inner clip area instead to position clip-path).
+    final String svgClipPath = elevation == 0.0 ?
+        _pathToSvgClipPath(path,
+            offsetX: -pathBounds.left,
+            offsetY: -pathBounds.top,
+            scaleX: 1.0 / pathBounds.width,
+            scaleY: 1.0 / pathBounds.height)
+        : _pathToSvgClipPath(path,
+            offsetX: 0.0,
+            offsetY: 0.0,
+            scaleX: 1.0 / pathBounds.right,
+            scaleY: 1.0 / pathBounds.bottom);
+    /// If apply is called multiple times (without update) , remove prior
+    /// svg clip and render elements.
     _clipElement?.remove();
+    _svgElement?.remove();
     _clipElement =
         html.Element.html(svgClipPath, treeSanitizer: _NullTreeSanitizer());
     domRenderer.append(rootElement!, _clipElement!);
+    if (elevation == 0.0) {
+      DomRenderer.setElementStyle(
+          rootElement!, 'clip-path', 'url(#svgClip$_clipIdCounter)');
+      DomRenderer.setElementStyle(
+          rootElement!, '-webkit-clip-path', 'url(#svgClip$_clipIdCounter)');
```

Comment:
Setting style properties has been showing up in our profiles. Can we set either `clip-path` or `-webkit-clip-path` depending on the `browserEngine`, but not both?

Reply:
done. Moved clip-path code into DomRenderer